File: src/LeetCode/LinkedList/Design/SinglyLinkedList.java
package LeetCode.LinkedList.Design;

public class SinglyLinkedList {
    public Node head;
    public Node tail;
    public int size;

    // Time complexity: O(1)
    // Space complexity: O(1)
    public Node createSinglyLinkedList(int nodeValue) {
        head = new Node();
        Node node = new Node();

        node.value = nodeValue;
        node.next = null;

        head = node;
        tail = node;
        size++;

        return head;
    }


    // Time complexity: O(n)
    // Space complexity: O(1)
    public void insert(int location, int nodeValue) {
        if (size == 0) {
            createSinglyLinkedList(nodeValue);
            return;
        }

        Node node = new Node();
        node.value = nodeValue;

        // beginning
        if (location == 0) {
            node.next = head;
            head = node;

            // end
        } else if (location >= size) {
            node.next = null;
            tail.next = node;
            tail = node;
            // middle
        } else {
            Node curr = head;

            for (int i = 0; i < location - 1; i++) {
                curr = curr.next;
            }

            Node nextNode = curr.next;
            curr.next = node;
            node.next = nextNode;
        }

        size++;
    }


    // Time complexity: O(n)
    // Space complexity: O(1)
    public void traversal() {
        if (head == null) {
            System.out.println("Please create singly linked list first!");
        } else {
            Node tempNode = head;

            for (int i = 0; i < size; i++) {
                System.out.print(tempNode.value);
                if (i != size - 1) {
                    System.out.print(" -> ");
                }

                tempNode = tempNode.next;

            }
            System.out.println();

        }


    }

    // Time complexity: O(n)
    // Space complexity: O(1)
    public boolean search(int nodeValue) {
        if (head != null) {
            Node tempNode = head;

            for (int i = 0; i < size; i++) {
                if (tempNode.value == nodeValue) {
                    System.out.println("Found at index " + i);
                    return true;
                }

                tempNode = tempNode.next;

            }

        }

        System.out.println("Not found");

        return false;
    }


    // Time complexity: O(n)
    // Space complexity: O(1)
    public void delete(int location) {
        if (head == null) {
            System.out.println("SLL not exist");
            return;
        }

        // deleting first element
        if (location == 0) {
            head = head.next;
            size--;

            if (size == 0) {
                head = null;
                tail = null;
            }


            // deleting last element
        } else if (location >= size) {
            Node temp = head;

            for (int i = 0; i < size - 1; i++) {
                temp = temp.next;
            }

            if (temp == head) {
                head = tail = null;
                size--;
                return;
            }

            temp.next = null;
            tail = temp;

            size--;

            // deleting the mid element
        } else {
            Node temp = head;

            for (int i = 0; i < location - 1; i++) {
                temp = temp.next;
            }

            temp.next = temp.next.next;

            size--;
        }

    }

    // Time complexity: O(1)
    // Space complexity: O(1)
    public void deleteEntireSLL() {
        head = null;
        tail = null;

        System.out.println("The SLL deleted successfully");
    }

}

File: src/LeetCode/typeScript/lib/SinglyLinkedList.ts
class MyNode {
  public value: number | null;
  public next: any;

  constructor(value: number | null = null) {
    this.value = value;
    this.next = null
  }

  print(): void {
    console.log("Value: ", this.value, " Next: ", this.next);
  }

}

class LinkedList {
  public head: MyNode
  public size: number;

  constructor() {
    this.head = new MyNode();
    this.size = 0;
  }


  get(index: number): number {
    if(this.head === null || index >= this.size) {
      return -1;
    }

    let tempNode: any = this.head;

    for(let i = 0; i < index; i++) {
      tempNode = tempNode.next;
    }

    return tempNode.value;
  }

  addAtIndex(index: number, value: number): void {
    if(index > this.size) {
      console.log("Index must be less than the size.")
      return;
    }

    if(index < 0) {
      index = 0;
    }

    if(!this.head.value) {
      this.head.value = value;

      this.size++;

      return;
    }

    const newNode = new MyNode(value);

    if(index === 0) {
      newNode.next = this.head;
      this.head = newNode;
    } else {
      let pred = this.head;

      for(let i = 0; i < index - 1; i++) {
        pred = pred.next;
      }

      newNode.next = pred.next;
      pred.next = newNode;
    }

    this.size++;
  }


  removeAtIndex(index: number): void {
    if(this.head === null) {
      console.log("Linked list is empty!");
      return;
    }

    if(index >= this.size) {
      console.log("The index must be less than the size.")
      return;
    }

    this.size--;

    if(index === 0) {
      this.head = this.head.next;
    } else {

      let pred = this.head;

      for(let i = 0; i < index - 1; i++) {
        pred = pred.next;
      }

      pred.next = pred.next.next;
    }
  }

  search(value: number): number {
    if(this.head === null) {
      console.log("LinkedList is empty!")
      
      return - 1;
    }

    let pred = this.head;

    for(let i = 0; i < this.size; i++) {
      if(pred.value === value) {
        console.log("The value found at index of: ", i);

        return pred.value;
      }

      pred = pred.next;
    }


    return -1;
  }

  traversal() {
    let tempNode = this.head;

    for(let i = 0; i < this.size; i++) {
      console.log(tempNode.value, " =>");

      tempNode = tempNode.next;
    }
  }


}


const ll = new LinkedList();

ll.addAtIndex(0, 1);
ll.addAtIndex(1, 2);
ll.addAtIndex(2, 3);
ll.traversal();

console.log("--------------------------------------------------------");
ll.traversal();

ll.removeAtIndex(3);
console.log("--------------------------------------------------------");
ll.traversal();

console.log("--------------------------------------------------------");
console.log(ll.search(5));
